File: Python/utilities/digitfinder.py
import re

def find_digits(value: str, allow_words: bool):
    digit_map = [
        [ "one", 1 ],
        [ "two", 2 ],
        [ "three", 3 ],
        [ "four", 4 ],
        [ "five", 5 ],
        [ "six", 6 ],
        [ "seven", 7 ],
        [ "eight", 8 ],
        [ "nine", 9 ]
    ]

    digits = []

    for i, ch in enumerate(value):
        if re.match("\\d", ch):
            digits.append(int(ch))
        if not allow_words:
            continue

        for word in digit_map:
            if value[i:i+len(word[0])] == word[0]:
                digits.append(word[1])
    return digits

File: CSharp/AdventOfCode.Core/StringUtilities/DigitFinder.cs
namespace AdventOfCode.Core.StringUtilities;

public static class DigitFinder
{
    public static IEnumerable<int> FindDigits(string input, bool allowWords)
    {
        for (var i = 0; i < input.Length; i++)
        {
            if (char.IsDigit(input[i]))
            {
                yield return int.Parse(input.Substring(i, 1));
                continue;
            }

            if (!allowWords)
                continue;

            foreach (var word in DigitWords)
            {
                if (i + word.Key.Length > input.Length)
                    continue;

                if (input.Substring(i, word.Key.Length).Equals(word.Key))
                {
                    yield return word.Value;
                }
            }
        }
    }

    private static readonly Dictionary<string, int> DigitWords = new()
    {
        { "one", 1 },
        { "two", 2 },
        { "three", 3 },
        { "four", 4 },
        { "five", 5 },
        { "six", 6 },
        { "seven", 7 },
        { "eight", 8 },
        { "nine", 9 }
    };
}
